src/main.rs

```diff
@@ -30,82 +31,126 @@ mod error;
 mod path;
 
 use midi::PortMidiNoteTracker;
+use pm::PortMidiDeviceId as DeviceId;
 use ui::Popup;
 use screen::*;
 use config::Config;
 use hardcode::*;
 use error::Result;
 
-fn print_devices(pm: &pm::PortMidi) {
-    for dev in pm.devices().unwrap() {
-        println!("{}", dev);
-    }
-}
-
 fn config_path() -> Result<PathBuf> {
     env::home_dir()
         .ok_or("Home directory not found".into())
         .map(|config_dir| config_dir.join(hardcode::CONFIG_FILE_NAME))
 }
 
-fn main() {
-    let context = pm::PortMidi::new().unwrap();
+type Looper = looper::Looper<PortMidiNoteTracker>;
 
-    let (input_id, output_id) = {
-        let args: Vec<String> = std::env::args().collect();
-
-        if args.len() < 2 {
-            print_devices(&context);
-            println!("Usage: ./dimooper <input-port> <output-port>");
-            std::process::exit(1);
-        }
-
-        (args[1].trim().parse().unwrap(), args[2].trim().parse().unwrap())
-    };
-
-    let in_info = context.device(input_id).unwrap();
-    println!("Listening on: {} {}", in_info.id(), in_info.name());
-    let in_port = context.input_port(in_info, 1024).unwrap();
-
-    let out_info = context.device(output_id).unwrap();
+fn create_looper(context: &pm::PortMidi, output_id: DeviceId) -> Result<Looper> { 
```

Comment:
This entire function suggests that `PortMidiNoteTracker` should have additional constructor that constructs the tracker from `pm::PortMidi` and `DeviceId`.

**This remark doesn't block the merge. Feel free to not fix it. I'm gonna file it as a techdebt to fix later.**